### ISSUE TITLE
A lazy audio context singleton

### ISSUE THREAD

**Issue by danigb** [audiojs/audio-context#6]
Hi @juliangruber !

I love the idea of your module. But I have a problem: I use `audio-context` mostly to make optional passing the audio context instance. For example `osc(440)` creates an oscillator with the default context, but you can use `osc(440, context)` to pass a different context.

The problem is that, if the user provides it own context, two context are created: the user one and the one from the `audio-context` module.

I know that this is a breaking change, but can you consider to wrap your singlenton inside a function?:

```js
var context = require('audio-context');
context().createOscillator();
```

or:

```js
var context = require('audio-context')();
context.createOscillator();
````

Thanks

**Comment by RangerMauve** [comment]
@danigb You could put the call to `require` within a function so that even though whatever you're using for the bundling will see the require in the source, you won't actually "activate" the module until you invoke the function that contains it.

E.g.

```javascript
function defaultContext(){
 return require("audio-context");
}

function yourInitialization(options){
 var context = options.context || defaultContext();
 context.createOscillator();
}
```

**Comment by juliangruber** [comment]
SGTM! Happy to accept a patch

**Comment by danigb** [comment]
Hi @RangerMauve, thanks for the suggestion. That's a possible solution with the advantage of not introduce a breaking change into the library 👍 

My only concern is, is that true for all environments? I seems to be true with node and browserify, but with other bundlers?

If so, I think it's better to leave the code as is, what do you think @juliangruber ?

**Comment by RangerMauve** [comment]
@danigb It won't work with anything that uses es6 imports since you can only use `import` statements at the top level of a module. (thus people can't wrap it in a function)

**Comment by dfcreative** [comment]
@danigb @juliangruber @RangerMauve should we go v1.0.0 with that improvement? I am :100: for.

**Comment by jamen** [comment]
Me too.  :smile:

**Comment by danigb** [comment]
Agree 👍

**Comment by dfcreative** [comment]
Do we still make it singleton:
```js
const createContext = require('audio-context')
let ctx1 = createContext()
let ctx2 = createContext()
ctx1 === ctx2
```
or we make it create new context every call, like [ios-safe-audio-context](https://github.com/jam3/ios-safe-audio-context)
```js
const createContext = require('audio-context')
let ctx1 = createContext()
let ctx2 = createContext()
ctx1 !== ctx2
```
or we use something in-between, like identifiers:
```js

const createContext = require('audio-context')
let ctx1 = createContext()
let ctx2 = createContext()
let ctx3 = createContext(someId)
ctx1 === ctx2
ctx1 !== ctx3
```
Personally I am for the third option.

@danigb @jamen @juliangruber ?

**Comment by juliangruber** [comment]
> const createContext = require('audio-context')
> let ctx1 = createContext()
> let ctx2 = createContext()
> ctx1 !== ctx2

this would defeat the purpose of this module as it wouldn't be a singleton any more.

If you need more contexts, I'd say just create them manually

**Comment by dfcreative** [comment]
Ok, possibly we have to consider `options`:
* **latencyHint** − see [MDN ref](https://developer.mozilla.org/en-US/docs/Web/API/AudioContext/AudioContext)
* **offline** − would be sweet to care less about inner package structure and just change it via `options`, kind of pattern in [webgl-context](https://www.npmjs.com/package/webgl-context), [2d-context](https://www.npmjs.com/package/2d-context), [get-canvas-context] and [`canvas.getContext`](https://developer.mozilla.org/en/docs/Web/API/HTMLCanvasElement/getContext)
* **channels**, **length** − actual params for [OfflineContext](https://developer.mozilla.org/en-US/docs/Web/API/OfflineAudioContext/OfflineAudioContext)
* **sampleRate** − according to [the spec](https://webaudio.github.io/web-audio-api/#idl-def-AudioContextOptions), that is valid option for any audio context.

So we cannot just simply have singleton, if context options differ. I'd consider Map or context IDs based on options. Like

```js
let ctx1 = createContext({sampleRate: 44100})
let ctx2 = createContext({sampleRate: 44100})
let ctx3 = createContext({sampleRate: 192000, latencyHint: "balanced"})

ctx1 === ctx2
ctx1 !== ctx3
```

**Comment by dfcreative** [comment]
Ok, closed in v1.0.0.
